ComboboxControl: Normalize hyphen-like unicode characters to ASCII hyphens when matching search queries

markbiek opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
When searching, treat hyphen-like characters as an ASCII hyphen (%2D).

## Why?
This is the implementation agreed upon in discussions on #41952. This improves the search but doesn't affect search result highlighting or cause any other accessibility regressions.

## How?
Following the [example outlined](https://github.com/WordPress/gutenberg/pull/41952#pullrequestreview-1037737092) by @ciampo , we have a list of hyphen-like characters that are replaced with an ASCII `-` when looking for search matches.

## Testing Instructions
Run
```
npm run test:unit ./packages/components/src/utils/test/strings.js
```
to verify string normalization is functioning properly.

## Screenshots or screencast <!-- if applicable -->

![image](https://user-images.githubusercontent.com/917632/182648024-75798a4c-cb2b-423d-977c-f7a1e1bab980.png)

![image](https://user-images.githubusercontent.com/917632/182648041-9c0eb230-6d40-48d7-a6ab-cd1ddd8a6980.png)

markbiek committed:
Update the ComboBoxControl to use normalizeTextString to handle hyphen-like characters

mirka commented on packages/components/src/combobox-control/index.js:
```diff
@@ -33,6 +33,74 @@ import { useControlledValue } from '../utils/hooks';
+		'\ufe58',
+		// ﹣ (small hyphen-minus)
+		'\ufe63',
+		// － (fullwidth hyphen-minus)
+		'\uff0d',
+	].join( '' ) }]`,
+	'g'
+);
+
+const normalizeTextString = ( value ) => {
+	return removeAccents( value )
+		.toLocaleLowerCase()
+		.replace( ALL_UNICODE_DASH_CHARACTERS, '-' );
+};
```
Feel free to extract this stuff out to a utils file if you want.

ciampo replied:
Should we also add some unit tests for the string normalization process?

mirka commented on packages/components/src/combobox-control/index.js:
```diff
@@ -82,11 +150,9 @@ function ComboboxControl( {
 		const startsWithMatch = [];
 		const containsMatch = [];
-		const match = removeAccents( inputValue.toLocaleLowerCase() );
+		const match = normalizeTextString( inputValue );
```
Really tasteful refactor here, thank you 💛

mirka commented on packages/components/src/combobox-control/stories/index.js:
```diff
@@ -287,3 +287,39 @@ _default.args = {
 	allowReset: false,
 };
+
+const NormalizedHyphensComboboxControl = () => {
+	const [ value, setValue ] = useState( null );
```
Our Storybook is still quite a mixed bag, but we are moving in a direction where we try to highlight practical usage examples, rather than treating it as a place to keep testing fixtures. And as much as I like this story, I feel like this normalization feature is a bit too specific to dedicate a story to. How about we move it to a unit test, instead of a story? Doesn't have to be comprehensive, just some assertion of the expected normalization behavior like:

```js
it( 'should normalize hyphen-like characters to hyphens', () => {
  expect( normalizeTextString( 'foo~bar' ).toBe( 'foo-bar' );
};
```

I wish the `matchingSuggestions` would be refactored to a pure function so that whole logic could be unit tested, but that's for another occasion 😄

mirka reviewed (commented):
Thank you, this is working very nicely!

I just had some stylistic thoughts, but feature-wise I think this is good to go 👍 If you could also add a quick [changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md), that'd be great.

markbiek committed:
Add normalizeTextString function

This removes accents, sets to locale-specific lowercase, and replaced hyphen-like unicode characters with an ASCII hyphen

ciampo commented on packages/components/src/utils/strings.js:
```diff
@@ -0,0 +1,73 @@
+	'g'
+);
+
+// @ts-ignore
```
Best way here would be to change the extension to `.ts` and add type annotations to the function.

Alternatively, adding a JSDoc comment should suffice

```suggestion
/**
 * Normalizes a string by removing accents, converting to lowercase, and
 * swapping all types of hyphen-like characters with the standard "-" character.
 *
 * @param {string} value The string to normalize
 * @return {string} The normalized string
 */
```

markbiek replied:
I went ahead and switched it to a `.ts` file. Thanks!

mirka commented on packages/components/CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@
 
 ### Enhancements
 
+- `ComboboxControl`: Normalize hyphen-like characters to an ASCII hyphn ([#42942](https://github.com/WordPress/gutenberg/pull/42942))
```
```suggestion
- `ComboboxControl`: Normalize hyphen-like characters to an ASCII hyphen ([#42942](https://github.com/WordPress/gutenberg/pull/42942))
```

mirka committed:
Merge branch 'trunk' into improvement/normalize-hyphens

mirka reviewed (approved):
Looking good! Thank you so much for this solid improvement 🚀
